[BUG] Fix Prophet with logistic growth #1079

k1m190r opened the pull request:
#### Reference Issues/PRs
Fixes #1079 and #2444
Copy and paste of #1868 

#### What does this implement/fix? Explain your changes.
Adds the possibility to define the required "cap" and "floor" parameters for Prophet when its growth is set to logistic.

#### Any other comments?
Suggest closing old relevant issues and PRs.

##### For all contributions
- [x] Tests pass locally and [private github-action](https://github.com/k1m190r/sktime/actions/runs/2280446935)
- [x] The PR title starts with [BUG]

fkiraly reviewed (changes requested):
Looks good to me!

my main blocking change requests would be:

* @aiwalter's comments
* I think we need to do something in case the columns `"cap"` and `"floor"` already exist. An option would be to temporarily rename them to `"cap__n"` etc where `n` is the smallest positive integer that the string isn't already a column name

k1m190r commented:
@fkiraly if `cap` and `floor`  cols exist in `y` should the values from the columns be used as `growth_cap` and `growth_floor` params and removed from `y`? Since Prophet expects univariate.

k1m190r commented on sktime/forecasting/fbprophet.py:
```diff
@@ -126,6 +126,8 @@ def __init__(
         add_country_holidays=None,
         # Args of fbprophet
         growth="linear",
+        growth_floor=0.0,
+        growth_cap=None,
```
In `Prophet` judging by the [code here ](https://github.com/facebook/prophet/blob/b7d5ec448177b7ac2abf25cc3259cb55e4c44069/python/prophet/forecaster.py#L321)`floor` default is 0.0, `cap` has no default and must be specified.

docstring added

k1m190r commented on sktime/forecasting/fbprophet.py:
```diff
@@ -142,7 +151,7 @@ def __init__(
         stan_backend=None,
         verbose=0,
-        interval_width=0,
+        interval_width=0.80,
```
Synced this [default parameter ](https://github.com/facebook/prophet/blob/b7d5ec448177b7ac2abf25cc3259cb55e4c44069/python/prophet/forecaster.py#L93)value with Prophet.

k1m190r commented:
Thank you for the corrections gents. I have noticed another [BUG] while implementing this PR. Will open new issue #2621.

fkiraly commented:
>@fkiraly if `cap` and `floor`  cols exist in `y` should the values from the columns be used as `growth_cap` and `growth_floor` params and removed from `y`? Since Prophet expects univariate.

No, I think in no case should values from `y` columns be used as parameters. That would break the unified interface.

fkiraly commented on sktime/forecasting/fbprophet.py:
```diff
@@ -38,6 +38,13 @@ class Prophet(_ProphetAdapter):
+    growth_floor: float, array like, default=0
+        Growth saturation minimum value. If float floor is constant.
+        If array like - Variable floor (must be same length as y).
+    growth_cap: float, array like
```
should say what the default is

fkiraly commented:
See #2626 on the `interval_width` param. I think it can be removed?

k1m190r commented:
> * optional parameter `growth_cap` needs to be indicated as optional in the docstring. Conditionally non-optional parameter `growth_floor` should say it is non-optional if `growth="logistic"` in the docstring.

- [ ] `growth_cap` is conditionally optional if `growth="logistic"` and `growth_floor` optionality implied by default=0?

> * I think array-like `growth_cap` and `growth_floor` do not work? So, docstring should not mention them.

- [x] removed


> * we can't change existing default params without deprecation, but I also think `interval_width` is not used, so I'd leave it or remove it entirely

- [x] reverted to 0

fkiraly commented:
>  is conditionally optional if growth="logistic" and growth_floor optionality implied by default=0?

Well, docstring needs to make clear that `growth_cap` is usually optional.
Currently, a reader could think setting `growth_cap` is mandatory in all cases.

k1m190r commented:
```
growth_floor: float, default=0
    Growth saturation minimum value.

growth_cap: float, default=None
    Growth saturation maximum aka carrying capacity.
    Conditionally optional: must be specified if `growth="logistic"`.
```

Is this ok?

fkiraly commented:
> Is this ok?

Not sure - is `None` not ok for `growth_cap` if `growth="logistic"`?
Would `1` make sense as a sensible default?

I would also write "Used only if `growth="logistic"`" for `growth_floor`.

fkiraly commented:
@k1m190r, are you still working on this?

k1m190r commented:
```
    growth: str, default="linear"
        String 'linear' or 'logistic' to specify a linear or logistic
        trend. If 'logistic' specified float for 'growth_cap' must be provided.
    growth_floor: float, default=0
        Growth saturation minimum value.
    growth_cap: float, default=None
        Growth saturation maximum aka carrying capacity.
        Conditionally optional: must be float if 'growth="logistic"'.
```

Ok I think this would be it.
- if `growth="linear"`: both `growth_floor` and growth_cap` are optional.
- if `growth="logistic"`: `growth_floor` is still optional (default=0) but `growth_cap` is not optional and must be non None - float.

fkiraly commented:
makes sense. I would remove the "conditionally optional" in favour of a clearer formulation, since this does not state what the condition is.

Last line, I would change to:
```
mandatory (float) iff `growth="logistic"`, has no effect and is optional, otherwise (if `growth` is not `"logistic"`)
```

k1m190r commented:
Done.

```
    growth: str, default="linear"
        String 'linear' or 'logistic' to specify a linear or logistic
        trend. If 'logistic' specified float for 'growth_cap' must be provided.
    growth_floor: float, default=0
        Growth saturation minimum value.
        Used only if  `growth="logistic"`, has no effect otherwise
        (if `growth` is not `"logistic"`).
    growth_cap: float, default=None
        Growth saturation maximum aka carrying capacity.
        Mandatory (float) iff `growth="logistic"`, has no effect and is optional,
        otherwise (if `growth` is not `"logistic"`).
```

fkiraly reviewed (approved):
Thanks!
